### PR TITLE
updating the asana plugin

### PR DESCRIPTION
**Pull request by justin1364**
Let us know what you think here - if people are using this, I'm guessing this will break the plugin for them?  Is there any way to look for people using this and reach out to them?  We're happy to help.

We are expiring API Keys and will stop supporting them, probably by Jan 20th, 2016 - so we'll need to get that part changed anyway.

Also, the lookups by name were failing in workspaces with many projects - like ours - making the integration unusable.

I've never used coffeescript - so please advise if I did anything horribly wrong.

Thanks!

### CODE REVIEW

**Comment by duncanhewett** [comment]
Justin,

Thanks for submitting this.

We are currently in the process of revamping all our issue tracker plugins, including Asana, so the coffee script plugins in this repository will no longer be used.  Do you have a firm date for when API key support is being turned off?  We are expecting to release our new plugins in mid-late January which would mean we might be able to avoid updating the existing plugin if we can release it before you turn off API keys.

As you point out, switching from API keys to Personal Access Tokens will break the integration until each user updates their configuration.  We have a fairly large number of users with an Asana integration set up so it would be good to avoid this.

Looking at your API documentation we are able to request an OAuth token from an API key.  This would mean we are able to maintain the existing integrations transparently.

We are planning to move to OAuth for authentication in any case, although the full flow implementation won't be ready initially.  In the interim, we will switch to Personal Access Tokens for users setting up a new Asana integration.

You mentioned that lookups by name were failing - do you know where the failure was?  If it was a failure at our end (i.e. too short a timeout) then we would prefer just to fix that.  We are trying to avoid users having to enter internal IDs to set up integrations in general, so it would be good to keep the workspace & project fields as names if possible.

Thanks,
Duncan

**Comment by justin1364** [comment]
Hey Duncan,

It looks like we're going to stop generating API Keys on Jan 20th and then kill them in a few months: https://asana.com/developers/feed/api-key-timeline-update1  - so that probably gives us a bit more time and we shouldn't have to worry about breakages for existing users.

Let me know how I can help for v2.  The problem with using the projects endpoint is on our side (we are returning 400s), because for large lists of projects, we need to paginate through them.

One way I've gotten around this in other integrations is simply creating a default projects for users, instead of having them choose one.  It's slightly more limiting, but I'm guessing the vast majority of use cases would want a dedicated project for Bugsnag issues anyway.

Basically - let us know how we can help build/test this so that it works for big teams in your next iteration!  We'd also be thrilled to help you get OAuth setup here if that's on your roadmap as well.

Justin

**Comment by duncanhewett** [comment]
Justin,

Good news that the existing API keys will continue to work for a few months - that gives us a bit of breathing space!  We'll switch to Personal Access Token for new integration setups and use the existing API keys for existing integrations until we move everyone over to OAuth.

For the projects endpoint issue, would we have more luck if we used the typeahead endpoint (https://asana.com/developers/api-reference/typeaheads#query) to query for the project name or is that likely to suffer from the same problem?

I think we'd be reluctant to create a dedicated project for Bugsnag issues ourselves but it could be an option.

Thanks for your offer of help, we'll be sure to let you know.  Potentially it may be useful it we had a large number of projects to test with but I suppose we could just create them though the API?  Do you know roughly what project count we should aim to test with?

Duncan

**Comment by justin1364** [comment]
The typeahead endpoint would be a great solution here.  I'd love to see
that in action.  I'm not sure how large a workspace you'd need to start
seeing the issues we're seeing - but I'm happy to help you test if you need
it.  It's likely that this doesn't affect many people (we're the oldest
Asana organization and we have tons of cruft from testing).

It sounds like we can close this PR.  Definitely keep me in the loop if you
need help on next steps.

**Comment by duncanhewett** [comment]
We'll get the logo updated as well!

I'll close this PR.
